epochStart/metachain/systemSCs.go

```diff
@@ -148,6 +155,41 @@ func (s *systemSCProcessor) processWithNewFlags(
 	return nil
 }
 
+func (s *systemSCProcessor) unStakeNodesWithNotEnoughFundsWithStakingV4(
+	validatorsInfoMap state.ShardValidatorsInfoMapHandler,
+	epoch uint32,
+) error {
+	nodesToUnStake, mapOwnersKeys, err := s.stakingDataProvider.ComputeUnQualifiedNodes(validatorsInfoMap)
+	if err != nil {
+		return err
+	}
+
+	log.Debug("unStake nodes with not enough funds", "num", len(nodesToUnStake))
+	for _, blsKey := range nodesToUnStake {
+		log.Debug("unStake at end of epoch for node", "blsKey", blsKey)
+		err = s.unStakeOneNode(blsKey, epoch)
+		if err != nil {
+			return err
```

Comment:
Right, this would be pretty bad: execution would fail here and we wouldn't be able to also update delegation contracts (by calling `updateDelegationContracts`). This code is from legacy, but also updated for staking v4, so yes, it would lead the inconsistencies.

Reply:
It will fail only if we make mistake in code